packages/vm/src/evm/precompiles/05-modexp.ts

```diff
@@ -1,74 +1,77 @@
-    fac2 = x.muln(480)
-    return fac1.add(fac2).subn(199680)
+    fac1 = x ** BigInt(2) / BigInt(16)
+    fac2 = x * BigInt(480)
+    return fac1 + fac2 - BigInt(199680)
   }
 }
 
-function multComplexityEIP2565(x: BN): BN {
-  const words = x.addn(7).divn(8)
-  return words.mul(words)
+function multComplexityEIP2565(x: bigint): bigint {
+  const words = (x + BigInt(7)) / BigInt(8)
+  return words * words
 }
 
-function getAdjustedExponentLength(data: Buffer): BN {
+function getAdjustedExponentLength(data: Buffer): bigint {
   let expBytesStart
   try {
-    const baseLen = new BN(data.slice(0, 32)).toNumber()
-    expBytesStart = 96 + baseLen // 96 for base length, then exponent length, and modulus length, then baseLen for the base data, then exponent bytes start
+    const baseLen = bufferToBigInt(data.slice(0, 32))
+    expBytesStart = 96 + Number(baseLen) // 96 for base length, then exponent length, and modulus length, then baseLen for the base data, then exponent bytes start
   } catch (e: any) {
     expBytesStart = Number.MAX_SAFE_INTEGER - 32
   }
-  const expLen = new BN(data.slice(32, 64))
+  const expLen = bufferToBigInt(data.slice(32, 64))
   let firstExpBytes = Buffer.from(data.slice(expBytesStart, expBytesStart + 32)) // first word of the exponent data
   firstExpBytes = setLengthRight(firstExpBytes, 32) // reading past the data reads virtual zeros
-  let firstExpBN = new BN(firstExpBytes)
+  let firstExpBigInt = bufferToBigInt(firstExpBytes)
   let max32expLen = 0
-  if (expLen.ltn(32)) {
-    max32expLen = 32 - expLen.toNumber()
+  if (expLen < BigInt(32)) {
+    max32expLen = 32 - Number(expLen)
   }
-  firstExpBN = firstExpBN.shrn(8 * Math.max(max32expLen, 0))
+  firstExpBigInt = firstExpBigInt >> (BigInt(8) * BigInt(Math.max(max32expLen, 0)))
 
   let bitLen = -1
-  while (firstExpBN.gtn(0)) {
+  while (firstExpBigInt > BigInt(0)) {
     bitLen = bitLen + 1
-    firstExpBN = firstExpBN.ushrn(1)
+    firstExpBigInt = firstExpBigInt >> BigInt(1)
   }
-  let expLenMinus32OrZero = expLen.subn(32)
-  if (expLenMinus32OrZero.ltn(0)) {
-    expLenMinus32OrZero = new BN(0)
+  let expLenMinus32OrZero = expLen - BigInt(32)
+  if (expLenMinus32OrZero < BigInt(0)) {
+    expLenMinus32OrZero = BigInt(0)
   }
-  const eightTimesExpLenMinus32OrZero = expLenMinus32OrZero.muln(8)
-  const adjustedExpLen = eightTimesExpLenMinus32OrZero
+  const eightTimesExpLenMinus32OrZero = expLenMinus32OrZero * BigInt(8)
+  let adjustedExpLen = eightTimesExpLenMinus32OrZero
   if (bitLen > 0) {
-    adjustedExpLen.iaddn(bitLen)
+    adjustedExpLen += BigInt(bitLen)
   }
   return adjustedExpLen
 }
 
-function expmod(B: BN, E: BN, M: BN): BN {
-  if (E.isZero()) return new BN(1).mod(M)
-  // Red asserts M > 1
-  if (M.lten(1)) return new BN(0)
-  const red = BN.red(M)
-  const redB = B.toRed(red)
-  const res = redB.redPow(E)
-  return res.fromRed()
+export function expmod(a: bigint, power: bigint, modulo: bigint) {
+  if (power === BigInt(0)) {
```

Comment:
This is performance-concious function, I think BigInt(0) and BigInt(1) can be assigned to _0n, _1n to not do the type casts over and over.

Reply:
Just to clarify my understanding, are you suggesting we cache the `BigInt(0)`/`BigInt(1)` just for this function to maximize performance or do it across the VM?  I thought we had agreed to not bother with the `_0n`/`_1n` constants and stick with the BigInt constructor everywhere.  I'm happy to go back and reverse the previous thought for `_0n` if there is enough of a performance differential between caching it and referencing vs just `BigInt(0)` but I wasn't aware there was any material difference there.

Reply:
Wanted to do the merge now so follow-up work can be done, maybe this can be done along a follow-up PR since not so critical.

Reply:
@acolytec3 ah, totally forget, yes I guess the *n -> BigInt(*) conversion still needs to be done. 😬 (or might there be some other solution? The TypeScript compiler will likely not do this? 🤔

Just tested on performance, there is indeed some substantial difference (but in a very low range), so might be worth to do on occasions where there is a potentially great round of repetitions:

```shell
> console.time('1');for (let i=0; i<= 1000000; i++) { let num = BigInt(0) }; console.timeEnd('1');
1: 18.209ms
undefined
> console.time('1');const _0n = BigInt(0); for (let i=0; i<= 1000000; i++) { let num = _0n }; console.timeEnd('1');
1: 1.449ms
undefined
```

Reply:
Update: ah, sorry sorry sorry, just seeing that you already did. Thanks a lot!